docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,159 @@ the worker pool has a finite size and even with the full batching algorithm in t
 one requesting thread, and although CellLoader2 also spins up many more requesting threads on the Cassandra side,
 the Cassandra cluster was unable to actually have these therads all do work concurrently.
 
-### Live Migrations and the Coordination Service
+### Determining Which Transaction Service To Use
+
+Effectively, we want to define an internal schema version for AtlasDB, and implement a mechanism for performing schema
+transitions while ensuring nodes that want to commit transactions agree on the state of the world - or at least,
+agree sufficiently that they won't contradict one another.
+
+For a given start timestamp, we want to be able to decide what mechanism to use to find the commit timestamp (or,
+alternatively, if the transaction was aborted).
+
+#### Coordination Service
+
+We introduce a notion of a coordination service, which agrees on values being relevant or correct at a given timestamp.
+The sequence of values being agreed needs to evolve in a *backwards consistent* manner - that is, one should avoid
+changes that may affect behaviour for decisions made at timestamps before the point where one is making one's update.
+More formally, if we read a value for some timestamp ``TS``, then all future values written must then ensure that
+decisions made at ``TS`` would be done in a way consistent with our initial read.
+
+Values provided are valid up to a specific timestamp, called the validity bound. When a transform is applied, a fresh
+timestamp ``F`` is taken, and the new value will be written with a validity bound ``F + C`` for some constant ``C``.
+Going forward, the behaviour for updates up to the previous bound must be preserved (which means decisions would be the
+same whether other service nodes read the new value or not). Nodes must not make decisions based on values read with a
+validity bound less than their timestamp of interest.
+
+```java
+public interface CoordinationService<T> {
+    Optional<ValueAndBound<T>> getValueForTimestamp(long timestamp);
+
+    CheckAndSetResult<ValueAndBound<T>> tryTransformCurrentValue(Function<ValueAndBound<T>, T> transform);
+
+    // ONLY FOR METRICS AND MONITORING - NOT FOR PRODUCTION DECISIONS.
+    Optional<ValueAndBound<T>> getLastKnownLocalValue();
+}
+```
+
+#### Physical Implementation
+
+The coordination service needs to persist state; we've implemented this as a non-transactional table in the AtlasDB
+key-value service called ``_coordination``. Within this table, each sequence of values being agreed on takes up one row.
+The values themselves are stored as dynamic columns with a var-long dynamic column key.
+
+- The value stored at column 0 is a ``SequenceAndBound`` which indicates the sequence number of the currently valid
```

Comment:
somewhat ambiguous, you have `column1` and `column2`, there is no `column 0` and `value` is itself a column, so do you mean the value inside `column1` or do you mean the value inside the `value` column where `column1 == 0x00`?

It may make sense to talk about these in terms of triples i.e.

> consider the triple `(sequence_id, sequence_number, value)` to map to a physical table where:
> ```
> key = sequence_id
> column1 = sequence_number
> column2 = unused
> value = value
> ```

Then write everything in terms of the triples.

Either that or you have to be pretty explicit about what you mean when you say "value". 

NB: Not entirely clear whether those map to the "columns" but that's what I've inferred as I've found it quite confusing.

Reply:
At this point we haven't introduced the Cassandra terms yet (`column1` and `column2` are Cassandra-schema specific constants that are used across Atlas), so here I refer to the column 0 in a dynamic column sense (which physically in Cassandra would be the value in `value` where `column1 = 0x00`). The real physical part probably starts after this.

Your inference is correct - though note that we can't use `column1` and `column2` in general to talk about physical tables, because these are Cassandra-specific. I've added a section above the Cassandra table example recalling the correspondence between the Atlas and Cassandra layers for dynamic columns.

Reply:
I've separated this paragraph and some of the discussion later on (which isn't really about the physical layer) into a persistence section, and left the physical section with the table.